### PR TITLE
[Feature] Add Burst control capability for Applications

### PR DESCRIPTION
**Pull request by shilmyhasan**
The PR adds the capability to support burst control for Application policies similar to what the product already has for Subscription policies.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java [hunk 1]
```diff
@@ -627,6 +638,15 @@ private boolean doThrottle(MessageContext messageContext) {
             initThrottleForSubscriptionLevelSpikeArrest(messageContext, authenticationContext);
             context.stop();
         }
+
+        if(authenticationContext != null && authenticationContext.getApplicationSpikesArrestLimit() > 0) {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java:642]: format

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java [hunk 2]
```diff
@@ -833,6 +861,83 @@ private OMElement createSpikeArrestSubscriptionLevelPolicy(String policyName, in
...
+        OMElement parsedPolicy = null;
+        StringBuilder policy = new StringBuilder(APIThrottleConstants.WS_THROTTLE_POLICY_HEADER);
+        if (maxCount != 0 && unitTime != 0) {
+            policy.append(createPolicyForRole(policyName, Integer.toString(unitTime), Integer.toString(maxCount)));
+        }
+        policy.append(APIThrottleConstants.WS_THROTTLE_POLICY_BOTTOM);
+        try {
+            parsedPolicy = AXIOMUtil.stringToOM(policy.toString());
+        } catch (XMLStreamException e) {
+            log.error("Error occurred while creating policy file for Hard Throttling.", e);
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java:875]: add a reason why this exception is ignored (without throwing an exception and stopping the flow)

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java [hunk 3]
```diff
@@ -833,6 +861,83 @@ private OMElement createSpikeArrestSubscriptionLevelPolicy(String policyName, in
...
+        }
+        return parsedPolicy;
+    }
+    private void initThrottleForApplicationLevelSpikeArrest(MessageContext synCtx,
+                                                            AuthenticationContext authenticationContext) {
+        policyKeyApplication = authenticationContext.getApplicationTier();
+        String applicationLevelThrottleKey = authenticationContext.getApplicationId()+":"+authenticationContext.getUsername();
+        int maxRequestCount = authenticationContext.getApplicationSpikesArrestLimit();
+        int applicationSpikeArrestWindowUnitTime = 0;
+        if(maxRequestCount != 0) {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java:885]: format

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java [hunk 4]
```diff
@@ -833,6 +861,83 @@ private OMElement createSpikeArrestSubscriptionLevelPolicy(String policyName, in
...
+            parsedPolicy = AXIOMUtil.stringToOM(policy.toString());
+        } catch (XMLStreamException e) {
+            log.error("Error occurred while creating policy file for Hard Throttling.", e);
+        }
+        return parsedPolicy;
+    }
+    private void initThrottleForApplicationLevelSpikeArrest(MessageContext synCtx,
+                                                            AuthenticationContext authenticationContext) {
+        policyKeyApplication = authenticationContext.getApplicationTier();
+        String applicationLevelThrottleKey = authenticationContext.getApplicationId()+":"+authenticationContext.getUsername();
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java:882]: format

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 5]
```diff
@@ -12886,8 +12894,15 @@ private void setCommonParametersForPolicy(PreparedStatement policyStatement, Pol
         } else {
             policyStatement.setString(11, UUID.randomUUID().toString());
         }
+
     }
 
+    private void setRateLimitDetails(ApplicationPolicy policy, ResultSet resultSet) throws SQLException {
+        if(resultSet.getInt(ThrottlePolicyConstants.COLUMN_RATE_LIMIT_COUNT) > 0) {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:12901]: format

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/AuthenticationContext.java [hunk 6]
```diff
@@ -180,6 +180,25 @@ public void setConsumerKey(String consumerKey) {
         this.consumerKey = consumerKey;
     }
 
+    private int applicationSpikesArrestLimit;
```
Comment (Arshardh) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/AuthenticationContext.java:183]: Shall we add the variables along with the other variables on the top?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java (hunk 1 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java]: fixed in b56bf3020abe507022f90c1b46e64547b10cf7ac

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java (hunk 2 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java:876]: added in b56bf3020abe507022f90c1b46e64547b10cf7ac

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java (hunk 3 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java]: fixed in b56bf3020abe507022f90c1b46e64547b10cf7ac

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java (hunk 4 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/throttling/ThrottleHandler.java]: fixed in b56bf3020abe507022f90c1b46e64547b10cf7ac

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java (hunk 5 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: fixed in b56bf3020abe507022f90c1b46e64547b10cf7ac

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/AuthenticationContext.java (hunk 6 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/AuthenticationContext.java]: Refactored in https://github.com/wso2/carbon-apimgt/pull/12162/commits/547e395745eb20cdba4453df74dd33ff5a91ac7c

**Review by tharindu1st — APPROVED** [review]

**Review by chamilaadhi — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.common.jms/src/main/java/org/wso2/carbon/apimgt/common/jms/JMSTransportHandler.java [hunk 7]
```diff
@@ -50,7 +51,17 @@ public class JMSTransportHandler {
...
     private static final Object lock = new Object();
 
-    public JMSTransportHandler(Properties jmsConnectionProperties) {
+    /**
+     * Constructs a JMSTransportHandler with specified JMS connection properties and JMS task manager properties.
+     *
+     * @param jmsConnectionProperties  The JMS connection properties to utilize.
+     *                                 If empty, default properties will be loaded.
+     * @param jmsTaskManagerProperties The properties for JMS task management.
+     *                                 If null, default values will be applied.
```
Comment (Arshardh) [inline components/apimgt/org.wso2.carbon.apimgt.common.jms/src/main/java/org/wso2/carbon/apimgt/common/jms/JMSTransportHandler.java:60]: We only have one param here right? We can remove the jmsConnectionProperties from the doc comment.

**Review by chamilaadhi — APPROVED** [review]

**Pull request merged**